### PR TITLE
Issues/reader content sanitization

### PR DESCRIPTION
**Pull request by aerych**
This PR takes the text sanitization and formatting code used by the ReaderPostServiceRemote, ports it to Swift and moves it into a separate helper utility.  Tests were updated accordingly. One thing to note is the video press formatting logic is removed with this PR.  Modern videopress is delivered as an iframe, not a collection of divs, script tags, and flash embeds and the check was obsolete. 

This sets the stage for comments to adopt the same sanitization in #6251. 

To Test:
Run the updated tests and ensure everything passes. 
Ensure that nothing looks amiss in the reader detail (e.g. no extra space between paragraphs)

Needs review: @kurzee, would you mind taking this one as its necessary for the comment refresh.

### CODE REVIEW

File: WordPress/WordPressTest/ReaderPostServiceRemoteTests.m [hunk 1]
```diff
@@ -126,27 +126,27 @@ - (void)testSummaryIsPlainText {
...
+//- (void)testRemoveForbiddenTags {
+//    ReaderPostServiceRemote *remoteService = nil;
+//    XCTAssertNoThrow(remoteService = [self service]);
+//
+//    NSString *str = @"<p>test</p><p>test</p>";
+//    NSString *styleStr = @"<script>alert();</script><style>body{color:#000;}</style><p>test</p><script>alert();</script><style>body{color:#000;}</style><p>test</p><script>alert();</script><style>body{color:#000;}</style>";
+//    NSString *sanitizedStr = [remoteService removeForbiddenTags:styleStr];
+//    XCTAssertTrue([str isEqualToString:sanitizedStr], @"The forbidden tags were not removed.");
+//
+//}
```
Comment (kurzee) [inline WordPress/WordPressTest/ReaderPostServiceRemoteTests.m]: Just double checking, did you want to leave these tests around commented out?

**Comment by kurzee** [comment]
Looking at this post: https://wordpress.com/read/blogs/53424024/posts/18844

It seems the spacing as rendered on WPiOS drops the horizontal line but keeps the content spacing. This may be way out of scope, but perhaps we can draw that line? Or maybe drop any extra spacing if we know where the line is supposed to be? No idea if that's even what's going on here.

**iPad:**
screen shot 2016-12-05 at 4 24 42 pm

**Web:**
screen shot 2016-12-05 at 4 26 13 pm

**Comment by kurzee** [comment]
The line spacing for block quotes seems to be a bit higher than seen on the web. The paragraph spacing looks good though: https://wordpress.com/read/blogs/60462319/posts/14232

**iPad:**

screen shot 2016-12-05 at 4 29 02 pm

**Web:**

screen shot 2016-12-05 at 4 29 33 pm

**Comment by kurzee** [comment]
The line spacing of general paragraph spacing on the detail view seems larger than on post cards, which may be intentional.

**Card:**

screen shot 2016-12-05 at 4 32 13 pm

**Detail:**

screen shot 2016-12-05 at 4 33 15 pm

**Comment by kurzee** [comment]
The spacing between an image and a paragraph on this post seems to be larger above the image, than seen on the web: https://wordpress.com/read/blogs/1549125/posts/1992

**iPad:**

screen shot 2016-12-05 at 4 34 38 pm

**Web:**

screen shot 2016-12-05 at 4 34 51 pm

**Comment by kurzee** [comment]
Suuuuuuuuuper nitpicky, but the spacing above the image is a bit less than the spacing below the image between these two paragraphs for this post: https://wordpress.com/read/blogs/53424024/posts/19978

**iPad:**
screen shot 2016-12-05 at 4 38 23 pm

**Comment by kurzee** [comment]
Trying to view this post causes a crash: https://wordpress.com/read/blogs/1129682/posts/28647

1. Add a tag for "apple".
2. Scroll down a few posts to find the post.
3. Select it, crash

Note: iPad 2, iOS 10.1 Simulator

```
2016-12-05 16:42:56.352 WordPress[28691:5742229] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'url shouldn't be nil'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000108ecd34b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000010892e21e objc_exception_throw + 48
	2   CoreFoundation                      0x0000000108ed1442 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x000000010308fe4d -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
	4   WordPress                           0x00000001003e25dd -[WPTableImageSource fetchImageForURL:withSize:indexPath:isPrivate:] + 397
	5   WordPress                           0x00000001004b3e1a _TFC9WordPress17WPRichContentView18imageForAttachmentfCS_16WPTextAttachmentCS_15WPRichTextImage + 778
	6   WordPress                           0x00000001004b414a _TToFC9WordPress17WPRichContentView18imageForAttachmentfCS_16WPTextAttachmentCS_15WPRichTextImage + 58
	7   WordPress                           0x00000001004b2f04 _TFC9WordPress17WPRichContentView17attachmentManagerfTCS_23WPTextAttachmentManager17viewForAttachmentCS_16WPTextAttachment_GSqCSo6UIView_ + 164
	8   WordPress                           0x00000001004b2faf _TToFC9WordPress17WPRichContentView17attachmentManagerfTCS_23WPTextAttachmentManager17viewForAttachmentCS_16WPTextAttachment_GSqCSo6UIView_ + 79
	9   WordPress                           0x00000001004aa109 _TFFC9WordPress23WPTextAttachmentManager20enumerateAttachmentsFT_T_U_FTGSqPs9AnyObject__VSC8_NSRangeGSpV10ObjectiveC8ObjCBool__T_ + 601
	10  WordPress                           0x00000001004a9680 _TTRXFo_oGSqPs9AnyObject__dVSC8_NSRangedGSpV10ObjectiveC8ObjCBool__dT__XFdCb_dGSqPS___dS0_dGSpS2___dT__ + 80
	11  Foundation                          0x0000000103004b81 -[NSAttributedString enumerateAttribute:inRange:options:usingBlock:] + 278
	12  WordPress                           0x00000001004a9e07 _TFC9WordPress23WPTextAttachmentManager20enumerateAttachmentsfT_T_ + 887
	13  WordPress                           0x00000001004b1ec1 _TFC9WordPress17WPRichContentViewW14attributedTextGSQCSo18NSAttributedString_ + 65
	14  WordPress                           0x00000001004b1e65 _TFC9WordPress17WPRichContentViews14attributedTextGSQCSo18NSAttributedString_ + 229
	15  WordPress                           0x00000001004b1769 _TFC9WordPress17WPRichContentViews7contentSS + 1849
	16  WordPress                           0x000000010045f046 _TFC9WordPress26ReaderDetailViewControllerP33_6894C07EBB00C79B78761FE3A823FACE17configureRichTextfT_T_ + 790
	17  WordPress                           0x00000001004595ae _TFC9WordPress26ReaderDetailViewControllerP33_6894C07EBB00C79B78761FE3A823FACE13configureViewfT_T_ + 222
	18  WordPress                           0x0000000100454ffa _TFC9WordPress26ReaderDetailViewController11viewDidLoadfT_T_ + 762
	19  WordPress                           0x0000000100455032 _TToFC9WordPress26ReaderDetailViewController11viewDidLoadfT_T_ + 34
	20  UIKit                               0x0000000104069c99 -[UIViewController loadViewIfRequired] + 1258
	21  UIKit                               0x0000000104070102 -[UIViewController __viewWillAppear:] + 118
	22  UIKit                               0x000000010409afbf -[UINavigationController _startCustomTransition:] + 1290
	23  UIKit                               0x00000001040abc34 -[UINavigationController _startDeferredTransitionIfNeeded:] + 697
	24  UIKit                               0x00000001040acdc7 -[UINavigationController __viewWillLayoutSubviews] + 58
	25  UIKit                               0x00000001042a3d6f -[UILayoutContainerView layoutSubviews] + 223
	26  UIKit                               0x0000000103f8cf50 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1237
	27  QuartzCore                          0x0000000103a63cc4 -[CALayer layoutSublayers] + 146
	28  QuartzCore                          0x0000000103a57788 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	29  QuartzCore                          0x0000000103a57606 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	30  QuartzCore                          0x00000001039e5680 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 280
	31  QuartzCore                          0x0000000103a12767 _ZN2CA11Transaction6commitEv + 475
	32  UIKit                               0x0000000103ef3484 _afterCACommitHandler + 346
	33  CoreFoundation                      0x0000000108e71e17 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	34  CoreFoundation                      0x0000000108e71d87 __CFRunLoopDoObservers + 391
	35  CoreFoundation                      0x0000000108e56b9e __CFRunLoopRun + 1198
	36  CoreFoundation                      0x0000000108e56494 CFRunLoopRunSpecific + 420
	37  GraphicsServices                    0x000000010b900a6f GSEventRunModal + 161
	38  UIKit                               0x0000000103ec8964 UIApplicationMain + 159
	39  WordPress                           0x000000010016548e main + 62
	40  libdyld.dylib                       0x0000000109dfe68d start + 1
	41  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

**Comment by kurzee** [comment]
Interestingly, and maybe unrelated, when viewing the stream for the tag "apple", selecting a few of the posts display limited text:

simulator screen shot dec 5 2016 4 47 47 pm

But when viewed on the web, the post content shows a block and link for viewing the full post: https://wordpress.com/read/blogs/17146794/posts/183873

screen shot 2016-12-05 at 4 49 02 pm

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                                                 range: NSRange(location: 0, length: imgElementStr.characters.count))
+
+            mContent.replaceCharactersInRange(match.range, withString: mImageStr as String)
+        }
+
+        return mContent as String
+    }
+
+
+    /// Parses the specified string for the value of the specified attibute.
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `attibute`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    ///
+    /// - Paramaters:
+    ///     - attribute: The attribut whose value should be retrieved.
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `attribut`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            mContent.replaceCharactersInRange(match.range, withString: mImageStr as String)
+        }
+
+        return mContent as String
+    }
+
+
+    /// Parses the specified string for the value of the specified attibute.
+    ///
+    /// - Paramaters:
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `Paramaters`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+                                                                   range: NSRange(location: 0, length: content.characters.count),
+                                                                   withTemplate: "")
+
+        return content
+    }
+
+
+    /// Mutates gallery image URLs to be correctly sized.
+    ///
+    /// - Paramaters:
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `Paramaters`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                                                                   range: NSRange(location: 0, length: content.characters.count),
+                                                                   withTemplate: closePTag)
+
+        return content
+    }
+
+
+    /// Removes inline style attributes from the specified content string.
+    ///
+    /// - Paramaters:
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `Paramaters`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+                                                                   range: NSRange(location: 0, length: content.characters.count),
+                                                                   withTemplate: "")
+
+        return content
+    }
+
+
+    /// Converts DIV tags to P tags and removes duplicate or redundant tags.
+    ///
+    /// - Paramaters:
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `Paramaters`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        content = (content as NSString).stringByReplacingHTMLEmoticonsWithEmoji() as String
+        content = resizeGalleryImageURL(content, isPrivateSite: isPrivate)
+
+        return content
+    }
+
+
+    /// Removes forbidden HTML tags from the specified string.
+    ///
+    /// - Paramaters:
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `Paramaters`

File: WordPress/Classes/Utility/RichContentFormatter.swift [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+        // Gallery Images
+        static let galleryImgTags = try! NSRegularExpression(pattern: "<img[^>]*data-orig-file[^>]*/>", options: .CaseInsensitive)
+    }
+
+
+    /// Formats the specified content string for display. Forbidden HTML tags are
+    /// removed, paragraphs are normalized, etc.
+    ///
+    /// - Paramaters:
```
Comment (kurzee) [inline WordPress/Classes/Utility/RichContentFormatter.swift]: `Paramaters`

**Comment by aerych** [comment]
> It seems the spacing as rendered on WPiOS drops the horizontal line but keeps the content spacing

HRs are currently removed, but if they were wrapped in a paragraph we'll end up with the extra spacing.  Not really related to the methods affected here as that happens downstream from the remote.  I'd like to keep support for HRs by just drawing a line, but tbd. 

> The line spacing for block quotes seems to be a bit higher than seen on the web

I think this is coming from the use of Merriweather-Italic.  Maybe it would be better to just use Merriweather and ditch the italics. 🤔   Yeah, let's do that. 

> The line spacing of general paragraph spacing on the detail view seems larger than on post cards

Ooo. Good eye.  Unrelated to this change.  I think the card is correct and we need to update the line height used in the detail.

> The spacing between an image and a paragraph on this post seems to be larger above the image

At first glance, this appears to be due to the use of `figure` and `figcaption` tags wrapping the image and caption, and ```<p></p>``` inside the `figure` block.  That's something new we can filter.  I figure we'll find a few combinations like this over time.

> but the spacing above the image is a bit less than the spacing below the image

Nice catch. Not quite sure what's going on here, and I think its occurring with just about every image. Whatever is happening would most likely have something to do with `WPTextAttachment` and `WPTextAttachmentManager`.

> Trying to view this post causes a crash

Looks like the crash is due to trying to create an NSURL from a path containing encoded XML characters. The NSURL init returns nil and we fail on the assert. We can fix in a separate issue.

> Interestingly, and maybe unrelated, when viewing the stream for the tag "apple",

That looks like a new treatment for summary only posts.  Unrelated to the changes here.

**Comment by aerych** [comment]
Btw, its *awesome* that you wrote up all those :D

**Comment by aerych** [comment]
>  I think the card is correct and we need to update the line height used in the detail.

Looking closer its actually the opposite.

**Comment by kurzee** [comment]
>Btw, its awesome that you wrote up all those :D

I **love** these types of details 😃

**Comment by aerych** [comment]
Hiya @kurzee.  Thanks for the excellent review!  I've opened issues for everything except that [last issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/6271#issuecomment-265003462) with the apple story. I'll follow up with the squad and see what's going on with that and proceed from there.

Typos fixed. I obviously need new 👓 

Ready for another round.

**Comment by kurzee** [comment]
@aerych we're good to go! :shipit:

**Comment by aerych** [comment]
Thanks @kurzee !!!

**Pull request merged**